Added new decorator

sakshilucky25 opened the pull request:
Added a new decorator saksag

- it allows dry run by skipping method execution of the decorated method.
- it will work for all methods with varying parameters list. 
- it will return a user defined value as a response if method execution is skipped to mock the response.
- Corresponding test cases have been added.


Example 1:
        >>> @saksag(return_value=25)
        ... def foo(bar, baz=None):
        ...     return 42
        ...
        >>> a = foo("sport", baz=False)
        >>> print(a)
        42
        >>> dryrun(True)
        >>> b = foo("sport", baz=False)
        INFO:drypy.saksag:[DRYRUN] call to 'foo(sport, baz=False)'
        >>> print(b)
        25

    Example 2:
        >>> @saksag(return_value="I am Sakshi")
        ... def sakfoo(bar):
        ...     return "I am sakfoo"
        ...
        >>> a = sakfoo("sport")
        >>> print(a)
        'I am sakfoo'
        >>> dryrun(True)
        >>> b = sakfoo("sport")
        INFO:drypy.saksag:[DRYRUN] call to 'sakfoo(sport)'
        >>> print(b)
        'I am Sakshi'

sakagraw committed:
Added new decorator saksag for returning user defined value from the function

sakshilucky25 committed:
Merge pull request #1 from sakshilucky25/sakshilucky25/saksag_decorator

Added new decorator saksag for returning user defined value from the …

dzanotelli commented:
Hey, your code seems fine, and the use case too.
Still:
- the name of the new decorator is not. `saksag` means nothing. Something like `bounty` (to keep the far west theme) would be much better.
- missing sphinx documentation (needed, to upload it later on readthedocs)

sakagraw committed:
renamed decorator and added sphinx doc

sakshilucky25 committed:
Merge pull request #2 from sakshilucky25/sakshilucky25/saksag_decorator

Renamed decorator to sentinel and added sphinx doc

sakagraw committed:
typo

sakshilucky25 committed:
Merge pull request #3 from sakshilucky25/sakshilucky25/saksag_decorator

typo

sakshilucky25 commented:
@dzanotelli, thanks for the quick review.
1. Renamed the decorator to sentinel.
2. Added sphinx document. Please let me know if something is still missing. 

I have few queries too:
1. As I understand it, the dryrun flag controlling the dryrun mode in this library is a global flag. If I use this library in my code, which runs in multiple parallel sessions, there could be an issue. For instance, if 3 users execute this code concurrently on a single machine, the global flag may not reflect the correct value (true/false) for all users, even if one user's execution updates it. 
e.g. all 3 users starts execution by setting in True. and USER1's flow is bit faster and it resets the flag to False which will be reflected to USER2 and USER3 whereas, for these users the flag should still be True. 
This could lead to concurrency issues. Could you confirm if my understanding is correct and suggest how this can be managed?

2. In setup.py, the version is fetched by get_version() method which essentially runs 'git describe' to pick the version. 
e.g. After committing that tag (git tag -a v1.0.4-1 -m "initial tag"), 'git describes' returns the tag value such as v1.0.4-1. But if there are other commits on the repo, 'git describe' will return value as +-n commits. eg. v1.0.4-1[last commit id] which fails the installation. Will it not be better to pick the latest tag instead, using 'git tag'?
Being said that, you may have to re-add the existing(v1.0.4) or add the new tag for version on the master repo of the library after this PR is merged to master.

dzanotelli commented:
@sakshilucky25 you are welcome, and thank you for contributing.

1. nice, `sentinel` seems ok
2. i checked the sphinx, overall good. There are just lesser things to change (e.g., using value `42` both as arg and as return value in your example may lead to confusion). But no worries, ill fix it while releasing the new pypi version of the package.

It's also missing the changelog entry (and some `## [next] - unreleased` section), but again, i can fix it later.

About your questions:
1. Yes dryrun flag was born as a global flag. My initial use case was just testing one-shot scripts, e.g., CLI-like tools adding the `--dry-run` option in the cmd line to run `dryrun(True)`. So, I actually never needed to run it with parallel sessions. If you have an actual use case where parallelism should be supported, can I ask you to elaborate a little more about it? Probably it's better to open a new issue (describe the scenario, expected usage, using what, threads/asyncio/other? etc.), and move there our discussion, tests and improvements.
2. the `get_version()` func is meant to help me in the pypi release process, so using it just with clean tags (semver, just digits and docs, without leading `v` or trailing other chars). While developing, you can install it using the git syntax if you need a specific commit, or just creating a symlink to your drypy folder in the root of your project. Concluding, thanks for the notice about tags, im going to check and fix them for sure before releasing.

dzanotelli commented:
@sakshilucky25 I should be able to merge the pull request soon and release a new pypi version. Best,

sakshilucky25 commented:
Thanks @dzanotelli. Really appreciate the effort. Please let me know the final stable version of the library once the new change is released and is available publicly. I need to use the stable version of the library with this new change in my code.

dzanotelli commented:
Hey, it has already been published as version 1.1.0 two days ago, as long as the documentation on readthedocs.org.
Enjoy! And thanks for contributing.
